### ISSUE TITLE
"When variable type is apparent" doesn't apply to array initialization

### ISSUE THREAD

**Issue by davkean** [dotnet/roslyn#26531]
This the same bug I filed https://github.com/dotnet/roslyn/issues/23907, but it stated that it was fixed.

**Version Used**: 
Version 15.8.0 Preview 1.0 [27628.0.d15.8]

**Steps to Reproduce**:

![image](https://user-images.githubusercontent.com/1103906/39463525-ccab474e-4d5b-11e8-8577-865f84347d2a.png)

``` C#

public class Foo
{
    public void Method()
    {
        var files = new string[0];
    }

}
```

**Expected Behavior**:

No message

**Actual Behavior**:

Message	IDE0008	Use explicit type instead of 'var'	ConsoleApp300	C:\Users\davkean\Source\Repos\ConsoleApp300\ConsoleApp300\Program.cs	6	Active

**Comment by sharwell** [comment]
This is by design. See #24114.

**Comment by jcouv** [comment]
Specifically, the preference for explicit types on built-in types dominates the preference for `var` when type is apparent.

**Comment by davkean** [comment]
"When type is apparent" is one of the options but it doesn't work.

**Comment by sharwell** [comment]
Do you have a suggestion for different wording that would more accurately reflect the behavior? For example:

> When type is apparent, except in cases where the type is a built-in type

**Comment by Neme12** [comment]
The current set of options is:
```
csharp_style_var_for_built_in_types		true/fale
csharp_style_var_when_type_is_apparent	true/false
csharp_style_var_elsewhere				true/false
```

I believe having these options instead would make sense and only increase the number o possible combinations from 8 to 9 while solving this issue:
```
csharp_style_var_for_built_in_types		always/never/when_type_is_apparent
csharp_style_var_elsewhere				always/never/when_type_is_apparent
```

**Comment by davkean** [comment]
Yes that makes sense. I'm trying to understand under what circumstances someone would choose the following (which is effectively what options are enforcing now):

```
csharp_style_var_for_built_in_types		never
csharp_style_var_elsewhere				when_type_is_apparent

```

I've never seen that style anywhere. Whereas the style I want enforced both project-system and CoreFx/CoreCLR enforce.

**Comment by sharwell** [comment]
It covers things like this: https://github.com/benaadams/corefx/blob/a29b8b80468c1694d8125f9d76e8c335c3b6ed5f/src/System.Collections/tests/Generic/List/List.Generic.Tests.netcoreapp.cs#L23

**Comment by Neme12** [comment]
Unfortunately my proposal can't be done in a backwards compatible way because it removes 2 combinations that are currently possible. I created a table mapping current options to the proposed ones:

```
built_in_types	when_type_is_apparent	elsewhere	|	built_in_types	elsewhere

false			false					false		|	never			never
false			false					true		|	NOT AVAILABLE
false			true					false		|	never			when_type_is_apparent
false			true					true		|	never			always
true			false					false		|	always			never
true			false					true		|	NOT AVAILABLE
true			true					false		|	always			when_type_is_apparent
true			true					true		|	always			always
```
It would no longer be possible to prefer explicit type when type is apparent but var when it is not apparent. This does seem like a bizarre preference, but even if it was OK to remove this, I don't know how it could be done with editrconfig to map these.

On the other hand, this proposal would add 3 new possible options:
```c#
	built_in_types			elsewhere

1.	when_type_is_apparent	never
2.	when_type_is_apparent	always
3.	when_type_is_apparent	when_type_is_apparent

```
where number 3 is the desired one. 2 seems reasonable too and 1 is rather bizarre too.

**Comment by Neme12** [comment]
@sharwell 
I don't think so. I would expect `0` to be classified as "not apparent". `var` would only be suggested if you type `var i = new int()`

**Comment by Neme12** [comment]
At least it was my understanding that "apparent" means "the type is explicit inside the expression". Maybe I'm wrong.

**Comment by davkean** [comment]
@Neme12 That makes sense, I was grouping the following together when I filed https://github.com/dotnet/roslyn/issues/23714, but I like your idea better.

``` C#
var i = new int[];
var i = 0;

```

**Comment by sharwell** [comment]
Reopening to see if the proposal from @Neme12 could provide a suitable replacement for current behavior without being overwhelming.

**Comment by sharwell** [comment]
@Neme12 The analysis provided comparing the two sets of options is good, but currently misses some of the behaviors. I'm specifically interested in understanding the following cases:

| Setting | Explicit type was used | `var` was used |
| --- | --- | --- |
| `always` | ❌ | ✔️ |
| `never` | ✔️ | ❌ |
| `when_type_is_apparent` | ❓ | ✔️ |

The current options can represent "allow when type is apparent, and disallow otherwise". I'm concerned the new options would subtly change that to "require when type is apparent, and disallow otherwise".

**Comment by JimBobSquarePants** [comment]
Observed behavior indicates that this seems to only trigger for built in types.

With the following config:
``` 
csharp_style_var_for_built_in_types = false:warning
csharp_style_var_elsewhere = false:warning
csharp_style_var_when_type_is_apparent = true:warning
```

Array initialization  of a built in type triggers IDE0008 .

``` c#
var buffer = new byte[n];
```

![image](https://user-images.githubusercontent.com/385879/72125550-1e651d00-33bd-11ea-8ba6-1e3c6fb2bb18.png)


However custom types do not trigger the error.

``` c#
var buffer = new MyCustomType[n];
```
![image](https://user-images.githubusercontent.com/385879/72125670-93385700-33bd-11ea-9441-b33273137842.png)

**Comment by sharwell** [comment]
I reviewed https://github.com/dotnet/roslyn/issues/26531#issuecomment-385825621 and https://github.com/dotnet/roslyn/issues/26531#issuecomment-385985344 again today with @mavasani. I believe my original concerns can now be addressed by setting the severity for the _diagnostic ID_ instead of setting it as part of the code style configuration. For example, the "allow when apparent, disallow otherwise" setting could be:

```ini
csharp_style_var_for_built_in_types = never
csharp_style_var_elsewhere = when_type_is_apparent
# IDE0007: Use implicit type
dotnet_diagnostic.IDE0007.severity = hidden
# IDE0008: Use explicit type
dotnet_diagnostic.IDE0008.severity = warning
```
